src/main/java/cz/cvut/kbss/analysis/service/FaultTreeRepositoryService.java

```diff
@@ -207,12 +207,17 @@ private void processBehavior(Behavior behavior, FaultEvent parentFaultEvent) thr
 
         if (!impairingBehaviors.isEmpty()) {
             for (Behavior impairingBehavior : impairingBehaviors) {
-                if(isVisited(impairingBehavior))
-                    continue;
+                if(isVisited(impairingBehavior) || isFailureModeCause(impairingBehavior)) continue;
                 faultEvents.add(processImpairingBehavior(impairingBehavior, parentFaultEvent));
             }
+        }
```

Comment:
Code similar to the one in setFaultEventTypes except the condition. If it makes sense implement `setFaultEventTypes(boolean isBasic, FaultEvent fEvent)` and use it here and line 245.